_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,448 @@
+  and we were curious how synchronous vs. asynchronous flavours of the Elasticsearch client compare.
+
+## The Service
+
+For a description of the service to implement & benchmark,
+[see its API Spec document available in the locations-rs repository](https://github.com/strohel/locations-rs/blob/master/api-spec.md).
+
+We benchmark its simplest `/city/v1/get` endpoint,
+whose work is to look up a `city` document by id in Elasticsearch,
+look up associated `region` document,
+and render response JSON.
+It uses internal in-memory cache for regions (there is just tens of them),
+but no cache for much more numerous cities.
+
+When fetching documents from Elasticsearch,
+all implementations instruct it to strip (at least) the largest property `geometry`.[^strip]
+I believe this significantly strains Elasticsearch
+(or rather transfers the strain from the microservice to it).
+
+[^strip]: To significantly reduce required bandwidth between microservice and Elasticsearch.
+    This property is only used in Elasticsearch-side filtering, not by the microservice.
+
+## The Benchmark
+
+### Recipe
+
+The benchmark consists of multiple identical "runs" for each Dockerised implementation,
+see [bench-3-impls.py](https://github.com/strohel/locations-rs/blob/master/bench-3-impls.py).
+The runs are named after the implementation and their ordinal number,
+thus `rs-actix-1` is the first run of the Rust (Actix) implementation,
+and `kt-ktor-3` is the third run of Kotlin (Ktor) implementation.
+
+Each run simulates the first couple of minutes of the microservice instance, see [test-image.py](https://github.com/strohel/locations-rs/blob/master/test-image.py).
+Steps of one run:
+1. Docker container is started, measuring the time until the service starts responding,
+2. a suite of HTTP checks is made ensuring correct responses and error handling behaviour,
+3. HTTP load is generated using [wrk](https://github.com/wg/wrk)
+   in a series of 10-second long steps with increasing concurrent connection count;
+   the metrics are collected from `wrk` and Docker API after each step:
+   - first 5 steps use just one concurrent connection, acting as a warm-up,
+   - each of the next 11 steps doubles concurrent connection count, i.e. continuing with 2, 4,
+     8, ..., 1024, 2048.[^conn-count]
+
+[^conn-count]: The top-end concurrent connection counts of 1024 and 2048 are too high to be practical.
+    They are included to test framework behaviour under uncontrolled traffic spike.
+
+Finally, all runs are plotted using awesome [Pygal](http://www.pygal.org/) library,
+see [render-tests.py](https://github.com/strohel/locations-rs/blob/master/render-tests.py).
+Different runs of a single implementation use the same colour,
+allowing for visual inspection of the variance between them.
+
+### Runtime Environment
+
+Docker is told to constrain resources available to the microservice containers,
+as in a typical Kubernetes cluster.
```

Comment:
https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-resource-requests-and-limits
> While your Kubernetes cluster might work fine without setting resource requests and limits, you will start running into stability issues as your teams and projects grow. Adding requests and limits to your Pods and Namespaces only takes a little extra effort, and can save you from running into many headaches down the line!

People always follow best practices, right? ;)

Reply:
The article is kinda funny. A person with a position called "Developer Advocate" says:
> This could be caused by a team spinning up more replicas than they need to artificially decrease latency (hey, it’s easier to spin up more copies than make your code more efficient!), ...

and
> Regardless of whether the issue is caused by a bad developer, bad code, or bad luck, ...

;-)

Reply:
aaah, damn I understood the sentense completely differently; like
> Docker constrains resources in the same way as they're constrained in typical k8s cluster.

and not

> We tell docker to constrain resources as is typically done in a k8s cluster.

Reply:
Aha, then it calls for rewording